### PR TITLE
Format command for REPL #1567

### PR DESCRIPTION
**Pull request by dominicmkennedy**
I have added the machinery to print strings with arbitrary expressions within the Futhark REPL. I have yet to add any options for formatting, my first thought is to add formatting code to `Util.Pretty`, but I am open to advice

### CODE REVIEW

File: src/Futhark/CLI/REPL.hs [hunk 1]
```diff
@@ -405,6 +417,41 @@ typeCommand = genTypeCommand parseExp T.checkExp $ \(ps, e) ->
...
+printCommand :: Command
+printCommand input = do
+  case parseForPrint input of
+    Left err -> liftIO $ T.putStrLn err
+    Right (strs, exps) -> do
+      prompt <- getPrompt
+      case partitionEithers $ map (parseExp prompt) exps of
+        ((SyntaxError _ err) : _, _) -> liftIO $ T.putStr err
+        (_, uncheckedExps) -> do
+          vals <- mapM onExp uncheckedExps
```
Comment (athas) [inline src/Futhark/CLI/REPL.hs:444]: I have a hunch that continuing to evaluate later expressions, after an early one has failed, would be confusing.  I suggest stopping at the first error.

**Comment by athas** [comment]
Neat, this is a good start for functionality we've wanted for a while.

* I think using Python-style formatting is the right call.  We don't need to implement [every feature](https://docs.python.org/3/library/string.html#formatspec) right now, but we should not go down a path that prevents us from adding features later.  Supporting brace-based interpolation, without anything else, is a pretty good way to begin, and useful on its own.

* Util.Pretty is supposed to extend (and wrap) the prettyprinting library.  I don't think it's a good fit for parsing format strings.  I don't have a strong vision for where else to do it.  Perhaps a new module is best, since we also want to use this functionality in `futhark literate`.  Maybe call it Futhark.Format?

* Parsing format strings should eventually be done in a slightly smarter way, so we can put the right source locations in the expressions.  But what you have will suffice for now.

File: src/Futhark/CLI/REPL.hs (hunk 1 above)
Comment (dominicmkennedy) [reply src/Futhark/CLI/REPL.hs:447]: That sounds like a good idea, I added another commit resolving that, and also providing a *slightly* better error message if there is an error in an expression.

**Comment by dominicmkennedy** [comment]
Putting all of the formatting into it's own module seems like a good idea. If it's alright with you I'm going to mark this PR as ready to merge, but I'd be willing to help out with a Futhark.Format module.

File: src/Futhark/CLI/REPL.hs [hunk 2]
```diff
@@ -483,6 +533,15 @@ Only one source file can be loaded at a time.  Using the :load command a
 second time will replace the previously loaded file.  It will also replace
 any declarations entered at the REPL.
 
+|]
+      )
+    ),
+    ( "print",
```
Comment (athas) [inline src/Futhark/CLI/REPL.hs:539]: Call the command "format" instead.

File: src/Futhark/CLI/REPL.hs [hunk 3]
```diff
@@ -3,6 +3,7 @@
 -- | @futhark repl@
 module Futhark.CLI.REPL (main) where
 
+import Control.Arrow (ArrowChoice (left))
```
Comment (athas) [inline src/Futhark/CLI/REPL.hs:6]: I think you can use `Data.Bifunctor.first` instead of this (and that is the common style in the compiler).

**Review by athas — CHANGES_REQUESTED** [review]

File: src/Futhark/CLI/REPL.hs (hunk 3 above)
Comment (dominicmkennedy) [reply src/Futhark/CLI/REPL.hs]: And done

**Pull request merged**